prow/tide/history/history.go

```diff
@@ -0,0 +1,159 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// FE should support:
+// - filter by repo (?branch) (?pr)
+// - filter by author
+// - filter by action
+// - filter by errored state
+// - filter 'before time t'
+// - no-filter. i.e. all actions tide has taken in order.
+
+//
+// ???	Should `History` use a single `*recordLog` instead of per pool?
+//				- This would ensure that the views that look at multiple pools are not 'missing' events in some pools.
+//				- This means many more recent records are saved
+//					- better for active repos so we can increase history time window
+//				- Also means that non-recent records are dropped...
+//					- bad for inactive repos as we may GC most/all the history.
+//
+
+// Mock out time for unit testing.
+var now = time.Now
+
+type History struct {
+	logs map[string]*recordLog
+	sync.Mutex
+
+	logSizeLimit int
+}
+
+type Record struct {
+	Time    time.Time `json:"time"`
+	Action  string    `json:"action"`
+	BaseSHA string    `json:"baseSHA,omitempty"`
+	Target  []PRMeta  `json:"target,omitempty"`
+	Err     string    `json:"err,omitempty"`
+}
+
+type PRMeta struct {
+	Num    int    `json:"num"`
+	Author string `json:"author"`
+	Title  string `json:"title"`
+	SHA    string `json:"SHA"`
+}
+
+func New(maxRecordsPerKey int) *History {
+	return &History{
+		logs:         make(map[string]*recordLog),
+		logSizeLimit: maxRecordsPerKey,
+	}
+}
+
+func (h *History) Record(poolKey, action, baseSHA, err string, targets []PRMeta) {
+	h.Lock()
+	defer h.Unlock()
+
+	if _, ok := h.logs[poolKey]; !ok {
+		h.logs[poolKey] = newRecordLog(h.logSizeLimit)
+	}
+	h.logs[poolKey].add(&Record{
+		Time:    now(),
+		Action:  action,
+		BaseSHA: baseSHA,
+		Target:  targets,
+		Err:     err,
+	})
+}
+
+func (h *History) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	b, err := json.Marshal(h.AllRecords())
+	if err != nil {
+		logrus.WithError(err).Error("Encoding YAML history.")
+		b = []byte("{}")
+	}
+	if _, err = w.Write(b); err != nil {
+		logrus.WithError(err).Error("Writing YAML history response.")
+	}
+}
+
+func (h *History) AllRecords() map[string][]*Record {
+	h.Lock()
+	defer h.Unlock()
+
+	res := make(map[string][]*Record, len(h.logs))
+	for key, log := range h.logs {
+		res[key] = log.toSlice()
+	}
+	return res
+}
+
+// recordLog is a space efficient, limited size, append only list.
+type recordLog struct {
```

Comment:
This is wild and whacky and I wonder if we need it? How expensive is the silly simple impl?

Reply:
```go
type tideController {
    history []log
    length int
}

func (c *tideController) addHIstory(item log) {
    if len(c.history) == c.length {
        c.history = append(c.history[1:len(c.history)-1], item)
    } else {
        c.history = append(c.history, item)
    }
}
```

Reply:
There's a bunch of math and `%` etc in your impl that I definitely needed to read a couple of times to figure out what was going on

Reply:
Sure, if we think it will be static let's leave it alone. `tide` is complex enough that I'm having trouble touching it sanely so I think you can appreciate where I am coming from where the simpler (but more re-allocating) impl is five lines.

Reply:
Yeah I can definitely appreciate wanting to avoid complexity (determining if a job should run might be an undecidable problem 😄 ).
I think this is a pretty standard way to implement a sliding window though and once you add the slice reversal function and handle pools separately both choices are basically the same number of lines. 

I'll add a comment suggesting switching to the naive solution if this ever becomes an issue.

Reply:
fairly certain https://github.com/kubernetes/test-infra/pull/10530/files#r243660849 is going to do a lot of copy  and / or alloc one you fill the slice. 

glancing over the current code, this is just a fairly standard ring-buffer, I hope that we can maintain such a simple data structure :-) 

the only odd bit is the cached toSlice, and that's simple enough and should be cheap /shrug

one small optimization to consider in the future: currently we use `rl.limit` always, but we might want to consider the max of `cap(rl.q)`, `rl.limit` to avoid allocating unused space.

Reply:
> To read and validate
 ```go
    if len(c.history) == c.length {
        c.history = append(c.history[1:len(c.history)-1], item)
    }
```
> is trivial

FWIW, that should be `cap(c.history)` or the `c.limit` because otherwise we will only ever append a single item. Slices are `(data*, len, cap)`, `len` will be 0 initially after allocating.

We've validated what Cole has now and it should be fine and more efficient. I hope we can keep a rather large history around for users to look at, so this small complexity seems worth it.